ext/simplexml/simplexml.stub.php

```diff
@@ -10,69 +10,69 @@ function simplexml_import_dom(SimpleXMLElement|DOMNode $node, ?string $class_nam
 
     /**
-     * @return string|bool
+     * @tentative-return-type
      * @alias SimpleXMLElement::asXML
      */
-    public function saveXML(?string $filename = null) {}
+    public function saveXML(?string $filename = null): string|bool {}
 
-    /** @return array */
-    public function getNamespaces(bool $recursive = false) {}
+    /** @tentative-return-type */
+    public function getNamespaces(bool $recursive = false): array {}
 
-    /** @return array|false */
-    public function getDocNamespaces(bool $recursive = false, bool $fromRoot = true) {}
+    /** @tentative-return-type */
+    public function getDocNamespaces(bool $recursive = false, bool $fromRoot = true): array|false {}
 
-    /** @return SimpleXMLElement|null */
-    public function children(?string $namespaceOrPrefix = null, bool $isPrefix = false) {}
+    /** @tentative-return-type */
+    public function children(?string $namespaceOrPrefix = null, bool $isPrefix = false): ?SimpleXMLElement {}
 
-    /** @return SimpleXMLElement|null */
-    public function attributes(?string $namespaceOrPrefix = null, bool $isPrefix = false) {}
+    /** @tentative-return-type */
+    public function attributes(?string $namespaceOrPrefix = null, bool $isPrefix = false): ?SimpleXMLElement {}
 
     public function __construct(string $data, int $options = 0, bool $dataIsURL = false, string $namespaceOrPrefix = "", bool $isPrefix = false) {}
 
-    /** @return SimpleXMLElement|null */
-    public function addChild(string $qualifiedName, ?string $value = null, ?string $namespace = null) {}
+    /** @tentative-return-type */
+    public function addChild(string $qualifiedName, ?string $value = null, ?string $namespace = null): ?SimpleXMLElement {}
 
-    /** @return SimpleXMLElement|null */
-    public function addAttribute(string $qualifiedName, ?string $value = null, ?string $namespace = null) {}
+    /** @tentative-return-type */
+    public function addAttribute(string $qualifiedName, ?string $value = null, ?string $namespace = null): ?SimpleXMLElement {}
 
-    /** @return string */
-    public function getName() {}
+    /** @tentative-return-type */
+    public function getName(): string {}
 
     public function __toString(): string {}
 
-    /** @return int */
-    public function count() {}
+    /** @tentative-return-type */
+    public function count(): int {}
 
-    /** @return void */
-    public function rewind() {}
+    /** @tentative-return-type */
+    public function rewind(): void {}
 
-    /** @return bool */
-    public function valid() {}
+    /** @tentative-return-type */
+    public function valid(): bool {}
 
-    /** @return SimpleXMLElement|null */
-    public function current() {}
+    /** @tentative-return-type */
+    public function current(): ?SimpleXMLElement {}
 
-    /** @return string|false */
-    public function key() {}
+    /** @tentative-return-type */
+    public function key(): string|false {}
```

Comment:
Hm... so if the null/false values on current() and key() only occur when the iterator is used after it is exhausted, then I think we should throw `Error` in that case. Calling current/key on an exhausted iterator is a programming error, and I think we should avoid regressing iterator type information due to it.

Reply:
I'm not able to verify if these conditions only happen when the iterator is exhausted, but I implemented the suggested change. If you are right with it (e.g. the error message makes sense), and if it can be backported to 8.0 then I'll create a separate PR as well.